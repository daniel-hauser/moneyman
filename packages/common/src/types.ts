/**
 * Common types shared across packages
 */

import type { CompanyTypes } from "israeli-bank-scrapers";

/**
 * Type for images with captions, used for failure screenshots
 */
export type ImageWithCaption = {
  photoPath: string;
  caption: string;
};

/**
 * Type for run metadata
 */
export type RunMetadata = {
  domainsByCompany: Partial<Record<CompanyTypes | "infra", unknown>>;
  networkInfo: unknown;
  metadataLogEntries: Array<string>;
};
