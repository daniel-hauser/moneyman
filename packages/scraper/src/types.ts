import type { CompanyTypes } from "israeli-bank-scrapers";
import type { Transaction } from "israeli-bank-scrapers/lib/transactions.js";
import {
  type ScraperCredentials,
  type ScraperScrapingResult,
} from "israeli-bank-scrapers/lib/scrapers/interface.js";
export type { Transaction };

export type AccountConfig = ScraperCredentials & {
  companyId: CompanyTypes;
};

export interface TransactionRow extends Transaction {
  account: string;
  companyId: CompanyTypes;
  hash: string;
  uniqueId: string;
}

export interface AccountScrapeResult {
  companyId: CompanyTypes;
  result: ScraperScrapingResult;
}

export type ScraperConfig = {
  startDate: Date;
  futureMonthsToScrape: number;
  parallelScrapers: number;
  accounts: Array<AccountConfig>;
};
