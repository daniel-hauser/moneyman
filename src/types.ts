import type { CompanyTypes } from "israeli-bank-scrapers";
import type { Transaction } from "israeli-bank-scrapers/lib/transactions.js";
import type {
  ScraperCredentials,
  ScraperScrapingResult,
} from "israeli-bank-scrapers";
import { SaveStats } from "./bot/saveStats.js";
export type { Transaction };

export type AccountConfig = ScraperCredentials & {
  companyId: CompanyTypes;
};

export interface TransactionRow extends Transaction {
  account: string;
  companyId: CompanyTypes;
  hash: string;
  uniqueId: string;
}

export interface AccountScrapeResult {
  companyId: CompanyTypes;
  result: ScraperScrapingResult;
}

export type CategoryString = string;
export type CategoryDef = {
  contains?: Array<string>;
  startsWith?: Array<string>;
  eq?: Array<string>;
};

export interface TransactionStorage {
  canSave(): boolean;
  saveTransactions(
    txns: Array<TransactionRow>,
    onProgress: (status: string) => Promise<void>,
  ): Promise<SaveStats>;
}

export type ScraperConfig = {
  startDate: Date;
  futureMonthsToScrape: number;
  parallelScrapers: number;
  accounts: Array<AccountConfig>;
  additionalTransactionInformation: boolean;
};

export type ImageWithCaption = {
  photoPath: string;
  caption: string;
};

export type RunMetadata = {
  domainsByCompany: Partial<Record<CompanyTypes, unknown>>;
  networkInfo: unknown;
  metadataLogEntries: Array<string>;
};

export interface RunnerHooks {
  onBeforeStart(): Promise<void>;
  onStatusChanged(rows: string[], totalTime?: number): Promise<void>;
  onResultsReady(results: AccountScrapeResult[]): Promise<void>;
  onError(e: Error, caller?: string): Promise<void>;
  failureScreenshotsHandler: (photos: ImageWithCaption[]) => Promise<unknown>;
  reportRunMetadata(metadata: RunMetadata): Promise<void>;
}
export type Runner = (hooks: RunnerHooks) => Promise<void>;
