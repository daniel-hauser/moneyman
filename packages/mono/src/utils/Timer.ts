export class Timer {
  duration: number;
  startTime: number;

  constructor(public name: string) {
    this.startTime = performance.now();
  }

  end() {
    this.duration = performance.now() - this.startTime;
  }

  toString() {
    if (this.duration === undefined) {
      return this.name;
    }
    return `${this.name}, took ${(this.duration / 1000).toFixed(2)}s`;
  }
}
