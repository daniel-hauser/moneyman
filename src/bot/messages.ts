import {
  TransactionStatuses,
  TransactionTypes,
} from "israeli-bank-scrapers/lib/transactions.js";
import { AccountScrapeResult, Transaction } from "../types.js";
import { normalizeCurrency } from "../utils/currency.js";
import { Timer } from "../utils/Timer.js";

export function getSummaryMessages(results: Array<AccountScrapeResult>) {
  const accountsSummary = results.flatMap(({ result, companyId }) => {
    if (!result.success) {
      return `\t‚ùå [${companyId}] ${result.errorType}${
        result.errorMessage ? `\n\t\t${result.errorMessage}` : ""
      }`;
    }
    return result.accounts?.map(
      (account) =>
        `\t‚úîÔ∏è [${companyId}] ${account.accountNumber}: ${account.txns.length}`,
    );
  });

  const { pending, completed } = transactionsByStatus(results);

  return `
${transactionsString(pending, completed)}

Accounts updated:
${accountsSummary.join("\n") || "\tüò∂ None"}

Pending txns:
${transactionList(pending) || "\tüò∂ None"}
`.trim();
}

function transactionsString(
  pending: Array<Transaction>,
  completed: Array<Transaction>,
) {
  const total = pending.length + completed.length;

  return `
${total} transactions scraped.
${total > 0 ? `(${pending.length} pending, ${completed.length} completed)` : ""}
${foreignTransactionsSummary(completed)}
`.trim();
}

function foreignTransactionsSummary(completed: Array<Transaction>) {
  const original = completed.filter(
    (tx) => normalizeCurrency(tx.originalCurrency) !== "ILS",
  ).length;

  if (original === 0) {
    return "";
  }

  const charged = completed.filter(
    (tx) => normalizeCurrency(tx.chargedCurrency) !== "ILS",
  ).length;
  return `From completed, ${original} not originally in ILS${
    charged ? ` and ${charged} not charged in ILS` : ""
  }`;
}

function transactionAmount(t: Transaction): number {
  switch (t.type) {
    case TransactionTypes.Normal:
      switch (t.status) {
        case TransactionStatuses.Pending:
          return t.originalAmount;
        case TransactionStatuses.Completed:
          return t.chargedAmount;
      }
    case TransactionTypes.Installments:
      return t.chargedAmount;
  }
}

function transactionString(t: Transaction) {
  const amount = transactionAmount(t);

  const sign = amount < 0 ? "-" : "+";
  const absAmount = Math.abs(amount).toFixed(2);

  return `${t?.description}:\t${sign}${absAmount}${
    t.originalCurrency === "ILS" ? "" : ` ${t.originalCurrency}`
  }`;
}

export function transactionList(
  transactions: Array<Transaction>,
  indent = "\t",
) {
  return transactions.map((t) => `${indent}${transactionString(t)}`).join("\n");
}

export function saving(storage: string, steps: Array<Timer> = []) {
  const stepsString = steps.map((s) => `\t${s}`).join("\n");
  return `üìù ${storage} Saving...\n${stepsString}`.trim();
}

function transactionsByStatus(results: Array<AccountScrapeResult>) {
  const allTxns = results
    .flatMap(({ result }) =>
      result.accounts?.flatMap((account) => account?.txns),
    )
    .filter((t): t is Transaction => t !== undefined);

  const pendingTxns = allTxns.filter(
    (t) => t.status === TransactionStatuses.Pending,
  );
  const scrapedTxns = allTxns.filter(
    (t) => t.status === TransactionStatuses.Completed,
  );

  return {
    pending: pendingTxns,
    completed: scrapedTxns,
  };
}
