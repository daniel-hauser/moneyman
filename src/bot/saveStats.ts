import { TransactionStatuses } from "israeli-bank-scrapers/lib/transactions.js";
import { TransactionRow } from "../types.js";
import { transactionList } from "./messages.js";
import { Timer } from "../utils/Timer.js";

export interface SaveStats {
  /**
   * Store name
   */
  name: string;
  /**
   * Store elements to be updated (Accounts, budgets, etc ...)
   */
  table: string | undefined;
  /**
   * Total scrapped transactions handled
   */
  total: number;
  /**
   * Newly added to store
   */
  added: number;
  /**
   * Total scrapped transactions that are pending status
   */
  pending: number;
  /**
   * Transactions that already exists in store and don't required an update
   */
  existing: number;
  /**
   * Transactions skipped for other reasons (missing accounts, ignored by API, etc.)
   */
  otherSkipped: number;
  /**
   * Scrapped transactions that are charged in foreign currency (not ILS)
   */
  highlightedTransactions?: Record<string, Array<TransactionRow>>;
}

/**
 * Calculate the number of skipped transactions (existing + pending + otherSkipped)
 * @param stats SaveStats object
 * @returns Total number of skipped transactions
 */
export function getSkippedCount(stats: SaveStats): number {
  return stats.existing + stats.pending + stats.otherSkipped;
}

/**
 * Generate skipped transactions string with breakdown
 * @param stats SaveStats object
 * @returns Formatted string for skipped transactions or empty string if none
 */
export function skippedString(stats: SaveStats): string {
  const skipped = getSkippedCount(stats);

  if (skipped === 0) {
    return "";
  }

  const parts: string[] = [];
  if (stats.existing > 0) parts.push(`${stats.existing} existing`);
  if (stats.pending > 0) parts.push(`${stats.pending} pending`);
  if (stats.otherSkipped > 0) parts.push(`${stats.otherSkipped} other`);

  return `${skipped} skipped (${parts.join(", ")})`;
}

/**
 * Create a new SaveStats object with the given name, table and transactions
 * @param name Store name
 * @param table Store elements to be updated (Accounts, budgets, etc ...)
 * @param transactions Scrapped transactions
 * @param stats Optional stats to be added to the new object
 */
export function createSaveStats<TInit extends Partial<SaveStats>>(
  name: string,
  table: string | undefined,
  transactions: Array<TransactionRow>,
  stats: TInit = {} as TInit,
): SaveStats & TInit {
  const total = transactions.length;
  const pending = transactions.filter(
    (tx) => tx.status === TransactionStatuses.Pending,
  ).length;

  return {
    name,
    table,
    total,
    added: 0,
    pending,
    existing: 0,
    otherSkipped: 0,
    ...stats,
  };
}

export function statsString(
  stats: SaveStats,
  saveDurationMs: number,
  steps: Array<Timer> = [],
): string {
  const header = `📝 ${stats.name}${stats.table ? ` (${stats.table})` : ""}`;
  const stepsString = steps.map((s) => `\t${s}`).join("\n");
  const skippedInfo = skippedString(stats);

  const skippedBreakdown = skippedInfo ? `\t${skippedInfo}\n` : "";

  return `
${header}${stepsString ? "\n" + stepsString : ""}
\t${stats.added} added${skippedBreakdown}
\ttook ${(saveDurationMs / 1000).toFixed(2)}s
${highlightedTransactionsString(stats.highlightedTransactions, 1)}`.trim();
}

export function highlightedTransactionsString(
  groups: Record<string, TransactionRow[]> | undefined,
  indent = 0,
) {
  if (!groups || Object.keys(groups).length === 0) {
    return "";
  }

  const indentString = "\t".repeat(indent);
  const groupsString = Object.entries(groups)
    .filter(([_, txns]) => txns.length > 0)
    .map(([name, txns]) => {
      const transactionsString = transactionList(txns, `${indentString}\t`);
      return `${indentString}${name}:\n${transactionsString}`;
    });

  if (groupsString.length === 0) {
    return "";
  }

  return `${indentString}${"-".repeat(5)}\n${groupsString}`;
}
