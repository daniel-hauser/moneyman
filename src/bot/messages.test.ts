import { CompanyTypes } from "israeli-bank-scrapers";
import { getSummaryMessages, saving } from "./messages.js";
import { AccountScrapeResult, Transaction, TransactionRow } from "../types.js";
import {
  TransactionStatuses,
  TransactionTypes,
} from "israeli-bank-scrapers/lib/transactions.js";
import { ScraperErrorTypes } from "israeli-bank-scrapers/lib/scrapers/errors.js";
import { createSaveStats, SaveStats, statsString } from "./saveStats.js";
import { Timer } from "../utils/Timer.js";

describe("messages", () => {
  describe("getSummaryMessages", () => {
    it("should return a summary message", () => {
      const txns1 = [
        transaction({
          chargedAmount: -20,
          originalAmount: -100,
          description: "ILS",
          chargedCurrency: "ILS",
          originalCurrency: "USD",
        }),
      ];

      const txns2 = [
        transaction({
          type: TransactionTypes.Installments,
        }),
        transaction({
          status: TransactionStatuses.Pending,
          originalAmount: 20,
        }),
        transaction({
          status: TransactionStatuses.Pending,
          originalAmount: 20,
          originalCurrency: "USD",
        }),
        transaction({
          status: TransactionStatuses.Pending,
          originalAmount: -20,
        }),
        transaction({
          description: "description2",
          memo: "memo2",
        }),
        transaction({
          chargedAmount: 20,
          originalAmount: 5,
          originalCurrency: "USD",
        }),
        transaction({
          chargedAmount: 1000,
          originalAmount: 1000,
          originalCurrency: "USD",
          chargedCurrency: "USD",
        }),
      ];
      const results: Array<AccountScrapeResult> = [
        {
          companyId: CompanyTypes.max,
          result: {
            success: true,
            accounts: [
              {
                accountNumber: "account1",
                txns: txns1,
              },
              {
                accountNumber: "account2",
                txns: txns2,
              },
            ],
          },
        },
      ];

      const stats: Array<SaveStats> = [
        createSaveStats("Storage 1", "TheTable", txns1 as any, {
          added: txns1.length,
        }),
        createSaveStats("Storage 2", "TheTable", txns2 as any, {
          added: txns2.length,
          highlightedTransactions: {
            Group1: [
              {
                account: "account1",
                companyId: CompanyTypes.max,
                hash: "hash1",
                type: TransactionTypes.Normal,
                date: new Date().toISOString(),
                processedDate: new Date().toISOString(),
                description: "description1",
                originalAmount: 10,
                uniqueId: "uniqueId1",
                originalCurrency: "ILS",
                chargedAmount: 10,
                status: TransactionStatuses.Completed,
              },
            ],
          },
        }),
      ];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map((stats) => statsString(stats, 350000));
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should return a summary message with no results", () => {
      const results: Array<AccountScrapeResult> = [];

      const stats: Array<SaveStats> = [];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map((stats) => statsString(stats, 0));
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should return a summary message with failed results", () => {
      const results: Array<AccountScrapeResult> = [
        {
          companyId: CompanyTypes.max,
          result: {
            success: false,
            errorType: ScraperErrorTypes.Generic,
            errorMessage: "Some error message",
          },
        },
        {
          companyId: CompanyTypes.hapoalim,
          result: {
            success: false,
            errorType: ScraperErrorTypes.ChangePassword,
          },
        },
        {
          companyId: CompanyTypes.hapoalim,
          result: {
            success: true,
            accounts: [
              {
                accountNumber: "account1",
                txns: [transaction({})],
              },
            ],
          },
        },
      ];

      const stats: Array<SaveStats> = [];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map((stats) => statsString(stats, 0));
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should return a summary message with installments", () => {
      const transactions = [
        transaction({
          type: TransactionTypes.Installments,
          chargedAmount: 20,
          originalAmount: 100,
          description: "should be +20",
        }),
        transaction({
          type: TransactionTypes.Installments,
          chargedAmount: -20,
          originalAmount: -100,
          description: "should be -20",
        }),
      ];

      const results: Array<AccountScrapeResult> = [
        {
          companyId: CompanyTypes.max,
          result: {
            success: true,
            accounts: [
              {
                accountNumber: "account1",
                txns: transactions,
              },
            ],
          },
        },
      ];

      const stats: Array<SaveStats> = [
        createSaveStats("Storage 1", "TheTable", transactions as any, {
          added: 2,
          highlightedTransactions: {
            SomeGroup: transactions.map<TransactionRow>((t) => ({
              account: "account1",
              companyId: CompanyTypes.max,
              hash: "hash1",
              uniqueId: "uniqueId1",
              ...t,
            })),
          },
        }),
      ];

      const summary = getSummaryMessages(results);
      expect(summary).toMatchSnapshot();

      const saveSummaries = stats.map((stats) => statsString(stats, 0));
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should not add empty groups", () => {
      const tx = {
        ...transaction({}),
        hash: "hash1",
        uniqueId: "uniqueId1",
        account: "account1",
        companyId: CompanyTypes.max,
      };
      const stats: Array<SaveStats> = [
        createSaveStats("Storage", "TheTable", [tx], {
          highlightedTransactions: {
            Group1: [],
            Group2: [
              {
                ...transaction({}),
                hash: "hash1",
                uniqueId: "uniqueId1",
                account: "account1",
                companyId: CompanyTypes.max,
              },
            ],
          },
        }),
      ];

      const saveSummaries = stats.map((stats) => statsString(stats, 0));
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should not add empty groups", () => {
      const stats: Array<SaveStats> = [
        createSaveStats("Storage", "TheTable", [], {
          highlightedTransactions: {
            Group1: [],
          },
        }),
      ];

      const saveSummaries = stats.map((stats) => statsString(stats, 0));
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should support steps", () => {
      const tx = {
        ...transaction({}),
        hash: "hash1",
        uniqueId: "uniqueId1",
        account: "account1",
        companyId: CompanyTypes.max,
      };
      const stats: Array<SaveStats> = [
        createSaveStats("Storage", "TheTable", [tx], {
          added: 1,
          highlightedTransactions: {
            Group1: [tx],
          },
        }),
      ];
      const steps: Array<Timer> = [
        Object.assign(new Timer("Step1"), { duration: 10 }),
        Object.assign(new Timer("Step2"), { duration: 100 }),
        Object.assign(new Timer("Step3"), { duration: 10_000 }),
        Object.assign(new Timer("Step4"), { duration: 100_456 }),
      ];
      const saveSummaries = stats.map((stats) => statsString(stats, 0, steps));
      expect(saveSummaries).toMatchSnapshot();
    });

    it("should support stats with skipped transactions", () => {
      const tx = {
        ...transaction({}),
        hash: "hash1",
        uniqueId: "uniqueId1",
        account: "account1",
        companyId: CompanyTypes.max,
      };
      const stats: Array<SaveStats> = [
        createSaveStats("Storage", "TheTable", [tx], {
          added: 1,
          skipped: 2,
          existing: 1,
          pending: 1,
          highlightedTransactions: {
            Group1: [tx],
          },
        }),
      ];
      const saveSummaries = stats.map((stats) => statsString(stats, 0));
      expect(saveSummaries).toMatchSnapshot();
    });
  });

  describe("saving", () => {
    it("should return a saving message", () => {
      const savingMessage = saving("Storage");
      expect(savingMessage).toMatchSnapshot();
    });

    it("should return a saving message with steps", () => {
      const steps: Array<Timer> = [
        Object.assign(new Timer("Step1"), { duration: 10 }),
        Object.assign(new Timer("Step2"), { duration: 100 }),
        Object.assign(new Timer("Step3"), { duration: 10_000 }),
        Object.assign(new Timer("Step4"), { duration: 100_456 }),
      ];
      const savingMessage = saving("Storage", steps);
      expect(savingMessage).toMatchSnapshot();
    });

    it("should return a saving message with not finished steps", () => {
      const savingMessage = saving("Storage", [
        Object.assign(new Timer("Step1"), { duration: 10 }),
        Object.assign(new Timer("Step2"), { duration: 100 }),
        Object.assign(new Timer("Step3"), { duration: 10000 }),
        new Timer("Step4"),
      ]);
      expect(savingMessage).toMatchSnapshot();
    });

    it("should return a saving message with one not finished step", () => {
      const savingMessage = saving("Storage", [new Timer("Step4")]);
      expect(savingMessage).toMatchSnapshot();
    });
  });
});

export function transaction(t: Partial<Transaction>): Transaction {
  return {
    type: TransactionTypes.Normal,
    date: new Date().toISOString(),
    processedDate: new Date().toISOString(),
    description: "description1",
    originalAmount: 10,
    originalCurrency: "ILS",
    chargedCurrency: "ILS",
    chargedAmount: t.status === TransactionStatuses.Pending ? 0 : 10,
    status: TransactionStatuses.Completed,
    ...t,
  };
}
