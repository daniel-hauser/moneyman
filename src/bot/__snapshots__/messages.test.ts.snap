// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messages getSummaryMessages should not add empty groups 1`] = `
[
  "📝 Storage (TheTable)
	0 added
	took 0.00s
	-----
	Group2:
		description1:	+10.00",
]
`;

exports[`messages getSummaryMessages should not add empty groups 2`] = `
[
  "📝 Storage (TheTable)
	0 added
	took 0.00s",
]
`;

exports[`messages getSummaryMessages should return a summary message 1`] = `
"8 transactions scraped.
(3 pending, 5 completed)
From completed, 3 not originally in ILS and 1 not charged in ILS

Accounts updated:
	✔️ [max] account1: 1
	✔️ [max] account2: 7

Pending txns:
	description1:	+20.00
	description1:	+20.00 USD
	description1:	-20.00"
`;

exports[`messages getSummaryMessages should return a summary message 2`] = `
[
  "📝 Storage 1 (TheTable)
	1 added
	took 350.00s",
  "📝 Storage 2 (TheTable)
	7 added	3 skipped (3 pending)

	took 350.00s
	-----
	Group1:
		description1:	+10.00",
]
`;

exports[`messages getSummaryMessages should return a summary message with failed results 1`] = `
"1 transactions scraped.
(0 pending, 1 completed)

Accounts updated:
	❌ [max] GENERIC
		Some error message
	❌ [hapoalim] CHANGE_PASSWORD
	✔️ [hapoalim] account1: 1

Pending txns:
	😶 None"
`;

exports[`messages getSummaryMessages should return a summary message with failed results 2`] = `[]`;

exports[`messages getSummaryMessages should return a summary message with installments 1`] = `
"2 transactions scraped.
(0 pending, 2 completed)

Accounts updated:
	✔️ [max] account1: 2

Pending txns:
	😶 None"
`;

exports[`messages getSummaryMessages should return a summary message with installments 2`] = `
[
  "📝 Storage 1 (TheTable)
	2 added
	took 0.00s
	-----
	SomeGroup:
		should be +20:	+20.00
		should be -20:	-20.00",
]
`;

exports[`messages getSummaryMessages should return a summary message with no results 1`] = `
"0 transactions scraped.

Accounts updated:
	😶 None

Pending txns:
	😶 None"
`;

exports[`messages getSummaryMessages should return a summary message with no results 2`] = `[]`;

exports[`messages getSummaryMessages should support stats with otherSkipped transactions 1`] = `
[
  "📝 Storage (TheTable)
	1 added	4 skipped (1 existing, 1 pending, 2 other)

	took 0.00s
	-----
	Group1:
		description1:	+10.00",
]
`;

exports[`messages getSummaryMessages should support stats with skipped transactions 1`] = `
[
  "📝 Storage (TheTable)
	1 added	2 skipped (1 existing, 1 pending)

	took 0.00s
	-----
	Group1:
		description1:	+10.00",
]
`;

exports[`messages getSummaryMessages should support steps 1`] = `
[
  "📝 Storage (TheTable)
	Step1, took 0.01s
	Step2, took 0.10s
	Step3, took 10.00s
	Step4, took 100.46s
	1 added
	took 0.00s
	-----
	Group1:
		description1:	+10.00",
]
`;

exports[`messages saving should return a saving message 1`] = `"📝 Storage Saving..."`;

exports[`messages saving should return a saving message with not finished steps 1`] = `
"📝 Storage Saving...
	Step1, took 0.01s
	Step2, took 0.10s
	Step3, took 10.00s
	Step4"
`;

exports[`messages saving should return a saving message with one not finished step 1`] = `
"📝 Storage Saving...
	Step4"
`;

exports[`messages saving should return a saving message with steps 1`] = `
"📝 Storage Saving...
	Step1, took 0.01s
	Step2, took 0.10s
	Step3, took 10.00s
	Step4, took 100.46s"
`;
